### DISCUSSION TITLE
How to use PooledConnectionProvider + UnifiedJedis + Transaction

### DISCUSSION THREAD

**Discussion by Datta19** [xetorthio/jedis#3485] [Q&A]
I want use PooledConnectionProvider + UnifiedJedis + Transaction but it throws exception, how can use this or is there any alternate way to use same

Code :

import java.time.Duration;
import org.apache.commons.pool2.PooledObjectFactory;
import org.apache.commons.pool2.impl.GenericObjectPoolConfig;

import redis.clients.jedis.Connection;
import redis.clients.jedis.ConnectionFactory;
import redis.clients.jedis.HostAndPort;
import redis.clients.jedis.JedisPooled;
import redis.clients.jedis.Transaction;
import redis.clients.jedis.UnifiedJedis;
import redis.clients.jedis.providers.PooledConnectionProvider;

public class TransactionTest {
	private static final PooledConnectionProvider connectionProvider = buildPoolConfig();

	private static PooledConnectionProvider buildPoolConfig() {
		HostAndPort hostAndPort = new HostAndPort("localhost", 6379);

		PooledObjectFactory<Connection> factory = new ConnectionFactory(hostAndPort);

		GenericObjectPoolConfig<Connection> poolConfig = new GenericObjectPoolConfig<Connection>();

		poolConfig.setMaxTotal(5);
		poolConfig.setMaxIdle(2);
		poolConfig.setMinIdle(2);
		poolConfig.setTestOnBorrow(true);
		poolConfig.setTestOnReturn(true);
		poolConfig.setTestWhileIdle(true);
		poolConfig.setMinEvictableIdleTime(Duration.ofSeconds(60));
		poolConfig.setTimeBetweenEvictionRuns(Duration.ofSeconds(30));
		poolConfig.setNumTestsPerEvictionRun(3);
		poolConfig.setBlockWhenExhausted(true);

		return new PooledConnectionProvider(factory, poolConfig);

	}

	public static void test() {
		try (UnifiedJedis client = new UnifiedJedis(connectionProvider.getPool().getResource())) {
			Transaction tr = client.multi();
			tr.jsonSetLegacy(System.currentTimeMillis() + " - " + Thread.currentThread().getId() + "ABC", "Value");
			tr.exec();
		}
	}
	public static void main(String args[]) {
		test();
	}
}

Exception:

Exception in thread "main" java.lang.IllegalStateException: It is not allowed to create Pipeline from this class redis.clients.jedis.UnifiedJedis
	at redis.clients.jedis.UnifiedJedis.multi(UnifiedJedis.java:4696)
	at com.test.jedisTest.Jedis_Test.TransactionTest.test(TransactionTest.java:52)
	at com.test.jedisTest.Jedis_Test.TransactionTest.main(TransactionTest.java:58)

**Comment by sazzad16** [comment]
You can create Transaction directly from the connection.

```
public static void test() {
	try (Transaction tr = new Transaction(connectionProvider.getPool().getResource())) {
		tr.jsonSetLegacy(System.currentTimeMillis() + " - " + Thread.currentThread().getId() + "ABC", "Value");
		tr.exec();
	}
}
```

**Reply by Datta19** [reply]
@sazzad16
Yes it is working but we seen that if we had poolConfig.setMaxTotal(100) and if we try to insert more than 100 records then it get stuck after inserting 100 records in redisdb (we seen 100+ connection opened in redisinsight). so do we need to create connectionProvider for each transaction? please refer below code for your reference

public static void test() {
		for(int i=0;i<100;i++)
		{
			try (Transaction tr = new Transaction(connectionProvider.getPool().getResource())) {
				tr.jsonSetLegacy(System.currentTimeMillis() + " - " + Thread.currentThread().getId() + "ABC", "Value");
				tr.exec();
			}
		}
	
	}
	public static void main(String args[]) {
		for(int i=0;i<5;i++)
		{
			Thread thread = new Thread(new Runnable() {
				@Override
				public void run() {	
					test();
				}
			});
			thread.start();	
		}
		
	}

**Reply by sazzad16** [reply]
@Datta19 You have to close the connection object to return it back to the pool. You can do that separately; or do `new Transaction(connectionProvider.getPool().getResource(), true, true)` instead.

**Reply by Datta19** [reply]
@sazzad16 Thanks for the quick response and Its is working fine as per your suggestion. Thanks a lot again!!!
